working/0546-patterns/patterns.md

```diff
@@ -0,0 +1,923 @@
+
+**TODO: Should we loosen this to match against any Iterables by using
+`elementAt()`?**
+
+**TODO: Can the pattern have a type argument? If so, can that have a type
+pattern?**
+
+### Map matcher
+
+```
+mapMatcher ::= "{" mapMatcherEntries "}"
+
+mapMatcherEntries ::= mapMatcherEntry ( "," mapMatcherEntry )* ","?
+mapMatcherEntry ::= expression ":" matcher
+```
+
+Matches and destructures maps. If the value's type does not implement `Map<K,V>`
+for some K and V, then the match fails. Otherwise, tests the entry patterns:
+
+1.  For each `mapMatcherEntry`:
+
+    1.  Evaluate `expression` and pass to `containsKey()` on the value.
+
+    1.  If this returns `false`, the map does not match.
+
+    1.  Otherwise, pass the expression's value to the value's index operator.
+        Match the entry's matcher against the resulting value. If it does not
+        match, the map does not match.
+
+1.  If all entries match, the map matches.
+
+Note that, unlike with lists, a matched map may have additional entries that
+are not checked by the matcher.
+
+If it is a compile-time error if any of the entry key expressions are not
+constant expressions.
+
+**TODO: Can the pattern have type arguments? If so, can they have type
+patterns?**
+
+### Typed matcher
+
+```
+typedMatcher ::= type identifier
+```
+
+Performs both a type test and a variable binding. Matches if the runtime type of
+the value is a subtype of `type`. If so, binds a new variable named `identifier`
+to the value with static type `type`. Unlike the corresponding binder pattern,
+this *can* "downcast" in that it checks the type first.
+
+**TODO: Allow `final`?**
+
+**TODO: Allow type pattern.**
+
+### Bind matcher
+
+```
+bindMatcher ::= ( "var" | "final" ) binder
+```
+
+A bind matcher lets a matching pattern also perform variable binding. By using
+bind matchers as subpatterns of a larger matched pattern, a single composite
+pattern can validate some condition and then bind one or more variables only
+when that condition holds.
+
+The `var` or `final` keyword indicates that a bind matcher is desired, followed
+by the binder pattern.
+
+A bind matcher always succeeds. Upon matching, it evaluates its binder using the
+matched value. If the bind matcher uses `var`, any variables created by the
+inner binders are assignable, otherwise they are final.
+
+### Type test matcher
+
+```
+typeTestMatcher ::= "is" type
+```
+
+A type test matcher can be used to match against the value's runtime type. The
+matcher succeeds if the value's runtime type is a subtype of `type`.
+
+**TODO: Define how this interacts with promotion.**
+
+**TODO: It's not clear if we want both this and typed matcher. We could allow a
+name after this to make the type check explicit as in `is int i`. Then a
+typedMatcher would disallow downcasting like the corresponding binder does. Or
+we could eliminate this and require users to do `int _` if they just want to
+test the type of the value and not bind.**
```

Comment:
I could be wrong, but my hunch is that we won't be able to support patterns in assignments. There's just too much potential syntax clash there. We can support them for variable declarations, of course because we can require some leading keyword like `var` or `final`, but a mutating assignment might be too hard.

Even if we do support patterns on the LHS of assignment expressions, that specific pattern wouldn't be supported because it's a matcher, not a binder. We'd only allow irrefutable binders, so you could do:

```dart
i as int = 42;
```

But not:

```dart
is int i = 3.2; // What happens?
```